hibernate-core/src/main/java/org/hibernate/dialect/SqlAstTranslatorWithMerge.java

```diff
@@ -207,21 +213,36 @@ protected void renderMergeInsert(OptionalTableUpdate optionalTableUpdate) {
 
 	protected void renderMergeDelete(OptionalTableUpdate optionalTableUpdate) {
 		final List<ColumnValueBinding> valueBindings = optionalTableUpdate.getValueBindings();
+		final List<ColumnValueBinding> optimisticLockBindings = optionalTableUpdate.getOptimisticLockBindings();
 
-		appendSql( " when matched " );
+		appendSql( " when matched" );
 		for ( int i = 0; i < valueBindings.size(); i++ ) {
 			final ColumnValueBinding binding = valueBindings.get( i );
 			appendSql( " and " );
 			binding.getColumnReference().appendReadExpression( this, "s" );
 			appendSql( " is null" );
 		}
+		for (int i = 0; i < optimisticLockBindings.size(); i++) {
+			final ColumnValueBinding binding = valueBindings.get( i );
+			appendSql(" and ");
+			binding.getColumnReference().appendColumnForWrite( this, null );
+			appendSql("=?");
```

Comment:
```suggestion
			binding.getColumnReference().appendColumnForWrite( this, "t" );
			appendSql( "=" );
			binding.getValueExpression().accept( this );
```

This ^^ will fix your problem with the parameter registration.  But the test still fails for me.  It *feels* like the row is simply not updated when the versioned predicate fails.

Reply:
```console

[Hibernate] 
    merge 
    into
        "UpsertVersionedTest$Record" as t 
    using (
        select
            cast(? as bigint) id,
            cast(? as varchar) message,
            cast(? as bigint) version
    ) as s 
        on (t.id=s.id) 
    when not matched 
        then insert (id, message, version) 
        values
            (s.id, s.message, s.version)  
    when matched 
    and t.version=? 
        then update set
            t.message=s.message,
            t.version=s.version
[subsystem] TRACE g.hibernate.orm.jdbc.bind JdbcBindingLogging:24 - binding parameter (1:BIGINT) <- [123]
[subsystem] TRACE g.hibernate.orm.jdbc.bind JdbcBindingLogging:24 - binding parameter (2:VARCHAR) <- [goodbye earth]
[subsystem] TRACE g.hibernate.orm.jdbc.bind JdbcBindingLogging:24 - binding parameter (3:BIGINT) <- [2]
[subsystem] TRACE g.hibernate.orm.jdbc.bind JdbcBindingLogging:24 - binding parameter (4:BIGINT) <- [1]

...


    select
        r1_0.id,
        r1_0.message,
        r1_0.version 
    from
        "UpsertVersionedTest$Record" r1_0 
    where
        r1_0.id=?
[subsystem] TRACE g.hibernate.orm.jdbc.bind JdbcBindingLogging:24 - binding parameter (1:BIGINT) <- [123]
[subsystem] TRACE ibernate.orm.jdbc.extract JdbcExtractingLogging:24 - extracted value (2:VARCHAR) -> [hello earth]
[subsystem] TRACE ibernate.orm.jdbc.extract JdbcExtractingLogging:24 - extracted value (3:BIGINT) -> [2]
```

Reply:
@sebersole I have added the `t` alias to the generated SQL as you suggest. But that doesn't help with the parameter binding.

Reply:
You missed the more important part of that suggestion which was the rendering of the "value"